service/mapper/transitions.go

```diff
@@ -97,52 +113,218 @@ func (t *Transitions) BootstrapState(s *State) error {
-	// following the first empty tree. This will ensure that we index all
-	// paths within the root tree.
-	checkpoint, err := t.load.Checkpoint()
+	// When bootstrapping, the loader injected into the mapper loads the root
+	// checkpoint.
+	tree, err := t.load.Trie()
 	if err != nil {
-		return fmt.Errorf("could not read checkpoint: %w", err)
-	}
-
-	// Here, we store all the paths, so we can index the payloads if wanted.
-	var paths []ledger.Path
-	if !t.cfg.SkipBootstrap {
-		paths = allPaths(checkpoint)
+		return fmt.Errorf("could not load root trie: %w", err)
 	}
-	s.forest.Save(checkpoint, paths, first)
+	paths := allPaths(tree)
+	s.forest.Save(tree, paths, first)
 
-	second := checkpoint.RootHash()
+	second := tree.RootHash()
 	t.log.Info().Uint64("height", s.height).Hex("commit", second[:]).Int("registers", len(paths)).Msg("added checkpoint tree to forest")
 
 	// We have successfully bootstrapped. However, no chain data for the root
 	// block has been indexed yet. This is why we "pretend" that we just
 	// forwarded the state to this height, so we go straight to the chain data
 	// indexing.
-	s.status = StatusForwarded
+	s.status = StatusUpdate
+	return nil
+}
+
+// ResumeIndexing resumes indexing the data from a previous run.
+func (t *Transitions) ResumeIndexing(s *State) error {
+	if s.status != StatusResume {
+		return fmt.Errorf("invalid status for resuming indexing (%s)", s.status)
+	}
+
+	// We need to know what the last indexed height was at the point we stopped
+	// indexing.
+	last, err := t.read.Last()
+	if err != nil {
+		return fmt.Errorf("could not get last height: %w", err)
+	}
+
+	// When resuming, the loader injected into the mapper rebuilds the trie from
+	// the paths and payloads stored in the index database.
+	tree, err := t.load.Trie()
+	if err != nil {
+		return fmt.Errorf("could not restore index trie: %w", err)
+	}
+
+	// After loading the trie, we should do a sanity check on its hash against
+	// the commit we indexed for it.
+	hash := flow.StateCommitment(tree.RootHash())
+	commit, err := t.read.Commit(last)
+	if err != nil {
+		return fmt.Errorf("could not get last commit: %w", err)
+	}
+	if hash != commit {
+		return fmt.Errorf("restored trie hash does not match last commit (hash: %x, commit: %x)", hash, commit)
+	}
+
+	// At this point, we can store the restored trie in our forest, as the trie
+	// for the last finalized block. We do not need to care about the parent
+	// state commitment or the paths, as they should not be used.
+	s.last = flow.DummyStateCommitment
+	s.next = commit
+	s.forest.Save(tree, nil, flow.DummyStateCommitment)
+
+	// Lastly, we just need to point to the next height. The chain indexing will
+	// then proceed with the first non-indexed block and forward the state
+	// commitments accordingly.
+	s.height = last + 1
+
+	// At this point, we should be able to start indexing the chain data for
+	// the next height.
+	s.status = StatusIndex
+	return nil
```

Comment:
```suggestion
	// At this point, we should be able to start indexing the chain data for
	// the next height.
	s.status = StatusIndex

	return nil
```

Reply:
My reasoning here is the same. This is part of the "pattern" of these functions, and changing the status is something that goes hand in hand with the return, so I like to keep it together. Especially since putting a new line between them for places where it's not the end of the function looks way more weird. Just keeping it consistent that way.